File: leetcode/24.py
# Definition for singly-linked list.
from pickle import LIST


class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def swapPairs(self, head: ListNode) -> ListNode:
        sentry = ListNode(0, head)
        head = sentry

        while head.next is not None and head.next.next is not None:
            tmp1 = head.next
            tmp2 = head.next.next
            head.next = tmp2
            tmp1.next = tmp2.next
            tmp2.next = tmp1
            head = tmp1

        return sentry.next


if __name__ == "__main__":
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
    print(Solution().swapPairs(head))

File: leetcode/24.c
#include <stdlib.h>

struct ListNode {
  int val;
  struct ListNode* next;
};

struct ListNode* swapPairs(struct ListNode* head) {
  struct ListNode sentry = {0, head};
  head = &sentry;

  struct ListNode* tmp1 = NULL;
  struct ListNode* tmp2 = NULL;
  while (head->next != NULL && head->next->next != NULL) {
    tmp1 = head->next;
    tmp2 = head->next->next;
    head->next = tmp2;
    tmp1->next = tmp2->next;
    tmp2->next = tmp1;
    head = tmp1;
  }

  return sentry.next;
}